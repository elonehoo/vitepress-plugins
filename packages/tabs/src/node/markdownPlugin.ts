import type MarkdownIt from 'markdown-it'
import { ruleBlockTabs } from './ruleBlockTabs'
import { parseTabsContent } from './parseTabsContent'

type Params = {
  shareStateKey: string | undefined
}

const parseParams = (input: string): Params => {
  if (!input.startsWith('=')) {
    return {
      shareStateKey: undefined
    }
  }

  const splitted = input.split('=')
  return {
    shareStateKey: splitted[1]
  }
}

export const tabsPlugin = (md: MarkdownIt) => {
  md.block.ruler.before('fence', '=tabs', ruleBlockTabs, {
    alt: ['paragraph', 'reference', 'blockquote', 'list']
  })
  md.renderer.rules.tabs = (tokens, index, _options, env) => {
    const token = tokens[index]

    const tabs = parseTabsContent(token.content)
    const renderedTabs = tabs.map(tab => ({
      label: tab.label,
      content: md.render(tab.content, env)
    }))

    const params = parseParams(token.info)

    const tabLabelsProp = `:tabLabels="${md.utils.escapeHtml(
      JSON.stringify(tabs.map(tab => tab.label))
    )}"`
    const shareStateKeyProp = params.shareStateKey
      ? `sharedStateKey="${md.utils.escapeHtml(params.shareStateKey)}"`
      : ''
    const slots = renderedTabs.map(
      tab => `<template #${tab.label}>${tab.content}</template>`
    )

    return `<PluginTabs ${tabLabelsProp} ${shareStateKeyProp}>${slots.join(
      ''
    )}</PluginTabs>`
  }
}
