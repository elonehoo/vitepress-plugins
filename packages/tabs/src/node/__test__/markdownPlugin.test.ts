import MarkdownIt from 'markdown-it'
import { test, expect } from 'vitest'
import { tabsPlugin } from '../markdownPlugin'
import dedent from 'ts-dedent'

const setup = () => {
  const md = MarkdownIt()
  md.use(tabsPlugin)
  return md
}

test('render correctly', () => {
  const md = setup()
  const result = md.render(dedent`
    :::=tabs
    ::a
    a
    ::b
    b
    :::
  `)
  expect(result).toMatchSnapshot()
})
