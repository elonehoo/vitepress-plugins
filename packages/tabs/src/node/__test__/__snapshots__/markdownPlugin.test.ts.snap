// Vitest Snapshot v1

exports[`render correctly 1`] = `
"<PluginTabs :tabLabels=\\"[&quot;a&quot;,&quot;b&quot;]\\" ><template #a><p>a</p>
</template><template #b><p>b</p>
</template></PluginTabs>"
`;
